### ISSUE TITLE
[Feature] Support skipping bad grad updates

### ISSUE THREAD

**Issue by yzhangcs** [pytorch/torchtitan#916]
Hi, I'm wondering is it ok to support skipping inf/nan grads in cases of encountering bad data.
Some thing like this

```py
            if job_config.training.skip_nan_inf and (
                grad_norm.isnan() or grad_norm.isinf()
            ):
                logger.warning(
                    f"Skipping optimizer step - detected invalid gradient norm: {grad_norm:.4f}"
                )
                optimizers.zero_grad()
                train_state.skipped_step += 1
            else:
                optimizers.step()
            lr_schedulers.step()
```
very glad to create a PR if you think it's necessary cc @tianyu-l .

**Comment by fegin** [comment]
How do we know if nan/inf is not caused by bad training/modeling/hyper-parameters? Would it be better that the training should stop when encountering bad loss and let the model author further debug

**Comment by tianyu-l** [comment]
I have two comments:
1. I was going to say the same thing as @fegin pointed out -- it potentially can make it hard to debug with this option on. Could you share some scenarios where skipping is the proper behavior? In the case of bad data, maybe it's better for whoever using the bad data to put a warning and skipping in the code, rather than putting it in torchtitan.
2. One can achieve this without modifying `train.py`, by doing the check and skip in a customized optimizer, inside the [`step()` function call](https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/optimizer.py#L93).

**Comment by lessw2020** [comment]
just wanted to add that if we do want to support this, then longer term it may be a lot more performant to re-purpose Ke's CUDA kernel as the nan checker... that is much faster than a simple check written in PyTorch eager.  (cc @kwen2501 )
